lib/cli/src/commands/publish.rs

```diff
@@ -0,0 +1,679 @@
+    /// Skip validation of the uploaded package
+    #[clap(long)]
+    pub no_validate: bool,
+    /// Directory containing the `wasmer.toml` (defaults to current root dir)
+    #[clap(name = "PACKAGE_PATH")]
+    pub package_path: Option<String>,
+}
+
+#[derive(Debug, Error)]
+enum PublishError {
+    #[error("Cannot publish without a module.")]
+    NoModule,
+    #[error("Unable to publish the \"{module}\" module because \"{}\" is not a file", path.display())]
+    SourceMustBeFile { module: String, path: PathBuf },
+    #[error("Unable to load the bindings for \"{module}\" because \"{}\" doesn't exist", path.display())]
+    MissingBindings { module: String, path: PathBuf },
+    #[error("Error building package when parsing module \"{0}\": {1}.")]
+    ErrorBuildingPackage(String, io::Error),
+    #[error(
+        "Path \"{0}\", specified in the manifest as part of the package file system does not exist.",
+    )]
+    MissingManifestFsPath(String),
+    #[error("When processing the package filesystem, found path \"{0}\" which is not a directory")]
+    PackageFileSystemEntryMustBeDirectory(String),
+}
+
+impl Publish {
+    /// Executes `wasmer publish`
+    pub fn execute(&self) -> Result<(), anyhow::Error> {
+        let mut builder = Builder::new(Vec::new());
+
+        let cwd = match self.package_path.as_ref() {
+            Some(s) => std::env::current_dir()?.join(s),
+            None => std::env::current_dir()?,
+        };
+
+        let manifest_path_buf = cwd.join("wasmer.toml");
+        let manifest = std::fs::read_to_string(&manifest_path_buf)
+            .map_err(|e| anyhow::anyhow!("could not find manifest: {e}"))
+            .with_context(|| anyhow::anyhow!("{}", manifest_path_buf.display()))?;
+        let mut manifest = wapm_toml::Manifest::parse(&manifest)?;
+        manifest.base_directory_path = cwd.clone();
+
+        if let Some(package_name) = self.package_name.as_ref() {
+            manifest.package.name = package_name.to_string();
+        }
+
+        if let Some(version) = self.version.as_ref() {
+            manifest.package.version = version.clone();
+        }
+
+        // See if a user is logged in. The backend should check for authorization on uploading
+        let (registry, username) =
+            wasmer_registry::whoami(self.registry.as_deref(), self.token.as_deref()).with_context(
+                || {
+                    anyhow::anyhow!(
+                        "could not find username / registry for registry = {:?}, token = {}",
+                        self.registry,
+                        self.token.as_deref().unwrap_or_default()
+                    )
+                },
+            )?;
+
+        let registry_present =
+            wasmer_registry::test_if_registry_present(&registry).unwrap_or(false);
+
+        if !registry_present {
+            return Err(anyhow::anyhow!(
+                "registry {} is currently unavailable",
+                registry
+            ));
+        }
```

Comment:
This should happen here. Basically, in the same call when getting the signed url is where we can check if the registry is available or not, otherwise we hammer the GraphQL backend

Reply:
Done in https://github.com/wasmerio/wasmer/commit/46a2d28f30de284ddef3bff417c0faa6b7e627dc